indico_toolkit/indico_wrapper/modelop.py

```diff
@@ -0,0 +1,131 @@
+from indico import IndicoClient
+from indico.queries import GraphQLRequest
+import json
+import string
+
+
+class ModelOp:
+
+    """
+    Class to assit in Model Group Operation calls.
+    """
+
+    def __init__(self, client: IndicoClient):
+        self.client = client
+        # self.task_type = None
+        self.model_settings = None
+        self.model_list = None
+
+    def get_model_settings(self, model_group_id: int, model_id: int = None) -> dict:
+        """
+        Returns options of most recent model within a model group as a dictionary. If a specific model id is passed, that corresponding options dictionary is returned.
+        model_list attribute contains all available models as a list. model_settings contains most recent model's training options
+        Args:
+            model_group_id (int): id of model group
+            model_id (int, optional): argument to return a specific model within a model group
+        Returns:
+            dict: dictionary of all model options, including: 
+            "id", "domain, ""high_quality", "interlabeler_resolution",
+            "sampling_strategy", "seed", "test_split", "weight_by_class_frequency",
+            "word_predictor_strength", and "model_training_options"
+
+        """
+
+        query = """
+  query getMoonbowModelGroupModels($modelGroupId: Int!) {
+    modelGroup(modelGroupId: $modelGroupId) {
+      id
+      models {
+        id
+        modelOptions
+      }
+    }
+  }
+                """
+        results = self.client.call(
+            GraphQLRequest(query, {"modelGroupId": model_group_id})
+        )
+        models = results["modelGroup"]["models"]
+        self.model_list = models
+        self.model_settings = json.loads(models[0]["modelOptions"])[
+            "model_training_options"
+        ]
```

Comment:
Important point here: some of the information we might be interested in as a caller isn't returned to us; it's set as an attribute on the object. And the method does three things: gets a list of models for a model ID, gets the model training options for the most recent model, and returns the specified or most recent model. That makes this function harder to reason about as the user ("How do I use this to get the information I need?") and as the developer ("How does this method get used? Can I refactor/change/improve it without breaking how it's used?").

Ideally we want to write pure functions that don't have side effects. Sometimes we need to modify the shared state on an object or global, but in this case the attributes aren't used by anything else. They just represent different information the caller might be interested in.

I'd recommend separating this into at least two methods. One to get all the model settings for a model group. And another to get the model settings for a specific model from a model group. (I think using the most recent model when no model ID is specified for the latter would be good behavior and easy enough to document and reason about. So no third method necessary).

```python
def get_model_options(
    self, model_group_id: int, model_id: int | None = None
) -> dict[str, object]:
    """
    Return model options for `model_id` of `model_group_id` 
    or the most recent model if `model_id` is not specified.
    """
    all_model_options = list(self.get_all_model_options(model_group_id))

    if model_id is None:
        return all_model_options[0]
    else:
        for model_options in all_model_options:
            if model_options["id"] == model_id:
                return model_options
        else:
            raise RuntimeError(
                f"Model group {model_group_id} does not have a model with ID {model_id}"
            )

def get_all_model_options(self, model_group_id: int) -> Iterator[dict[str, object]]:
    """
    Return model options for all models of `model_group_id`.
    """
    response = client.call(...)

    for model in response["modelGroup"]["models"]:
        model_options = json.loads(model["modelOptions"])
        model_options["id"] = model["id"]
        yield model_options
```